File: packages/ucm-schema/ts/src/degrade.ts
/**
 * Degradation engine.
 *
 * `degrade(ucm, channel)` returns the closest UCM the channel can render.
 *
 * Strategy (declarative, no surprises):
 *   1. If the channel supports the UCM as-is (validation passes), return it.
 *   2. If the channel does not support the capability, fall back to a `text`
 *      UCM whose body is `ucm.fallback_text`. This is the universal escape.
 *   3. If the channel supports the capability but a limit is exceeded:
 *        - quick_replies with too many buttons → degrade to `list` if
 *          `interactive.list` is supported; otherwise to `text` fallback.
 *        - list with too many rows → truncate to the channel limit and add
 *          a synthetic "ver más" row that maps to a text response.
 *        - everything else → degrade to `text` fallback.
 *   4. composite is degraded by recursively degrading each child; if the
 *      depth exceeds the channel limit, the composite is flattened.
 *
 * Degradation never throws. If we cannot produce something valid, we return
 * the text fallback, which every channel supports by contract (or it
 * shouldn't be a channel at all).
 */
import {
  UCMMessageSchema,
  UCM_VERSION,
  type UCMMessage,
} from "./types.js";
import {
  getChannel,
  channelSupports,
  inferCapabilities,
  type ChannelName,
  type ChannelProfile,
} from "./channels/capabilities.js";
import { validate } from "./validators/index.js";

export type DegradationStep = {
  reason: "capability" | "limit" | "depth";
  from: string;
  to: string;
  detail: string;
};

export type DegradeResult = {
  ucm: UCMMessage;
  changed: boolean;
  steps: DegradationStep[];
};

export function degrade(
  input: UCMMessage,
  channel: ChannelName | ChannelProfile,
): DegradeResult {
  const profile = typeof channel === "string" ? getChannel(channel) : channel;
  const steps: DegradationStep[] = [];
  const result = degradeInner(input, profile, steps);
  return {
    ucm: result,
    changed: steps.length > 0,
    steps,
  };
}

function degradeInner(
  ucm: UCMMessage,
  channel: ChannelProfile,
  steps: DegradationStep[],
): UCMMessage {
  // Capability check first — if the channel cannot render this type at all,
  // collapse to text fallback immediately.
  const caps = inferCapabilities(ucm.type, ucm.content as Record<string, unknown>);
  const unsupported = caps.filter((c) => !channelSupports(channel, c));
  if (unsupported.length > 0) {
    steps.push({
      reason: "capability",
      from: ucm.type,
      to: "text",
      detail: `channel "${channel.name}" lacks ${unsupported.join(", ")}`,
    });
    return toTextFallback(ucm);
  }

  // Composite: recurse first, then check depth.
  if (ucm.type === "composite") {
    const limit = channel.limits.compositeMaxDepth;
    if (limit != null && compositeDepth(ucm) > limit) {
      steps.push({
        reason: "depth",
        from: "composite",
        to: "composite (flattened)",
        detail: `depth > ${limit}`,
      });
      const flat = flattenComposite(ucm);
      const childResults = flat.content.children.map((c) =>
        degradeInner(c, channel, steps),
      );
      return { ...flat, content: { children: childResults } };
    }
    const childResults = ucm.content.children.map((c) =>
      degradeInner(c, channel, steps),
    );
    return { ...ucm, content: { children: childResults } };
  }

  // Structural limits: re-use validator and apply per-type fixes.
  const verdict = validate(ucm, channel);
  if (verdict.ok) return ucm;

  // Only proceed with limit-specific fixes; if there's a non-limit issue
  // (shape/capability that wasn't caught) we degrade to text safely.
  const onlyLimits = verdict.issues.every((i) => i.kind === "limit");
  if (!onlyLimits) {
    steps.push({
      reason: "capability",
      from: ucm.type,
      to: "text",
      detail: `validation failed: ${verdict.issues
        .map((i) => `${i.kind}@${i.path}: ${i.message}`)
        .join("; ")}`,
    });
    return toTextFallback(ucm);
  }

  switch (ucm.type) {
    case "quick_replies": {
      const max = channel.limits.quickRepliesMaxButtons ?? Infinity;
      const titleMax = channel.limits.quickRepliesTitleMaxChars ?? Infinity;
      const buttons = ucm.content.buttons.map((b) => ({
        id: b.id,
        title: truncate(b.title, titleMax),
      }));
      // Too many buttons: prefer `list` if available, else text fallback.
      if (buttons.length > max) {
        if (channel.capabilities.has("interactive.list")) {
          steps.push({
            reason: "limit",
            from: "quick_replies",
            to: "list",
            detail: `${buttons.length} buttons > ${max} — converted to list`,
          });
          const list: UCMMessage = {
            ucm_version: UCM_VERSION,
            message_id: `${ucm.message_id}::degraded`,
            type: "list",
            capabilities_required: ["interactive.list"],
            fallback_text: ucm.fallback_text,
            metadata: { ...ucm.metadata, degraded_from: "quick_replies" },
            content: {
              body: ucm.content.body,
              button_text: truncate(
                "Ver opciones",
                channel.limits.listButtonTextMaxChars ?? 20,
              ),
              sections: [
                {
                  title: truncate(
                    "Opciones",
                    channel.limits.listRowTitleMaxChars ?? 24,
                  ),
                  rows: buttons
                    .slice(0, channel.limits.listMaxRowsTotal ?? buttons.length)
                    .map((b) => ({
                      id: b.id,
                      title: truncate(
                        b.title,
                        channel.limits.listRowTitleMaxChars ?? 24,
                      ),
                    })),
                },
              ],
            },
          };
          return list;
        }
        steps.push({
          reason: "limit",
          from: "quick_replies",
          to: "text",
          detail: `${buttons.length} buttons > ${max} — channel cannot list, falling back to text`,
        });
        return toTextFallback(ucm);
      }
      // Truncated titles only.
      const someTruncated = buttons.some(
        (b, i) => b.title !== ucm.content.buttons[i]?.title,
      );
      if (someTruncated) {
        steps.push({
          reason: "limit",
          from: "quick_replies",
          to: "quick_replies (truncated)",
          detail: `button titles truncated to ${titleMax} chars`,
        });
        return { ...ucm, content: { ...ucm.content, buttons } };
      }
      return ucm;
    }
    case "list": {
      const maxRows = channel.limits.listMaxRowsTotal ?? Infinity;
      const rowTitleMax = channel.limits.listRowTitleMaxChars ?? Infinity;
      const rowDescMax = channel.limits.listRowDescriptionMaxChars ?? Infinity;
      const btnTextMax = channel.limits.listButtonTextMaxChars ?? Infinity;

      let count = 0;
      const sections = ucm.content.sections.map((s) => ({
        ...s,
        title: truncate(s.title, rowTitleMax),
        rows: s.rows
          .filter(() => {
            if (count >= maxRows) return false;
            count++;
            return true;
          })
          .map((r) => ({
            ...r,
            title: truncate(r.title, rowTitleMax),
            description:
              r.description != null ? truncate(r.description, rowDescMax) : r.description,
          })),
      }));
      const totalRowsOrig = ucm.content.sections.reduce(
        (acc, s) => acc + s.rows.length,
        0,
      );
      const truncatedRows = totalRowsOrig > maxRows;
      if (truncatedRows) {
        steps.push({
          reason: "limit",
          from: "list",
          to: "list (truncated)",
          detail: `${totalRowsOrig} rows > ${maxRows} — truncated`,
        });
      }
      const newButton = truncate(ucm.content.button_text, btnTextMax);
      if (newButton !== ucm.content.button_text) {
        steps.push({
          reason: "limit",
          from: "list",
          to: "list (truncated button_text)",
          detail: `button_text truncated to ${btnTextMax} chars`,
        });
      }
      return {
        ...ucm,
        content: { ...ucm.content, button_text: newButton, sections },
      };
    }
    case "cta_url": {
      const max = channel.limits.ctaUrlButtonTitleMaxChars ?? Infinity;
      const button_title = truncate(ucm.content.button_title, max);
      if (button_title !== ucm.content.button_title) {
        steps.push({
          reason: "limit",
          from: "cta_url",
          to: "cta_url (truncated button_title)",
          detail: `button_title truncated to ${max} chars`,
        });
      }
      return { ...ucm, content: { ...ucm.content, button_title } };
    }
    case "text":
    case "media": {
      const max = channel.limits.textBodyMaxChars ?? Infinity;
      if (ucm.type === "text") {
        const body = truncate(ucm.content.body, max);
        if (body !== ucm.content.body) {
          steps.push({
            reason: "limit",
            from: "text",
            to: "text (truncated)",
            detail: `body truncated to ${max} chars`,
          });
        }
        return { ...ucm, content: { ...ucm.content, body } };
      }
      // media with too-long caption
      if (ucm.content.caption != null) {
        const caption = truncate(ucm.content.caption, max);
        if (caption !== ucm.content.caption) {
          steps.push({
            reason: "limit",
            from: "media",
            to: "media (truncated caption)",
            detail: `caption truncated to ${max} chars`,
          });
        }
        return { ...ucm, content: { ...ucm.content, caption } };
      }
      return ucm;
    }
    default:
      return ucm;
  }
}

function toTextFallback(ucm: UCMMessage): UCMMessage {
  return {
    ucm_version: UCM_VERSION,
    message_id: `${ucm.message_id}::fallback`,
    type: "text",
    capabilities_required: ["text"],
    fallback_text: ucm.fallback_text,
    metadata: { ...ucm.metadata, degraded_from: ucm.type },
    content: { body: ucm.fallback_text, format: "plain" },
  };
}

function truncate(s: string, max: number): string {
  if (s.length <= max) return s;
  if (max <= 1) return s.slice(0, max);
  return s.slice(0, Math.max(1, max - 1)) + "…";
}

function compositeDepth(ucm: UCMMessage): number {
  if (ucm.type !== "composite") return 0;
  let m = 0;
  for (const c of ucm.content.children) {
    m = Math.max(m, compositeDepth(c));
  }
  return 1 + m;
}

function flattenComposite(ucm: UCMMessage): UCMMessage & { type: "composite" } {
  if (ucm.type !== "composite")
    throw new Error("flattenComposite called on non-composite");
  const out: UCMMessage[] = [];
  for (const child of ucm.content.children) {
    if (child.type === "composite") {
      out.push(...flattenComposite(child).content.children);
    } else {
      out.push(child);
    }
  }
  return { ...ucm, content: { children: out } };
}

File: packages/ucm-schema/py/src/ucm_schema/degrade.py
"""Channel-aware UCM degradation engine.

See `ts/src/degrade.ts` for the algorithm. The Python implementation is a
behavioural mirror; the cross-language test asserts equivalent outputs.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal

from .channels.capabilities import (
    ChannelName,
    ChannelProfile,
    channel_supports,
    get_channel,
    infer_capabilities,
)
from .types import (
    UCM_VERSION,
    CompositeUCM,
    ListContent,
    ListSection,
    ListRow,
    ListUCM,
    TextContent,
    TextUCM,
    UCMMessage,
)
from .validators import validate

Reason = Literal["capability", "limit", "depth"]


@dataclass(frozen=True)
class DegradationStep:
    reason: Reason
    from_type: str
    to_type: str
    detail: str


@dataclass(frozen=True)
class DegradeResult:
    ucm: UCMMessage
    changed: bool
    steps: list[DegradationStep] = field(default_factory=list)


def degrade(ucm: UCMMessage, channel_name: ChannelName) -> DegradeResult:
    channel = get_channel(channel_name)
    steps: list[DegradationStep] = []
    result = _degrade_inner(ucm, channel, steps)
    return DegradeResult(ucm=result, changed=bool(steps), steps=steps)


def _content_dict(ucm: UCMMessage) -> dict[str, Any]:
    return ucm.content.model_dump() if hasattr(ucm.content, "model_dump") else {}


def _degrade_inner(
    ucm: UCMMessage,
    channel: ChannelProfile,
    steps: list[DegradationStep],
) -> UCMMessage:
    caps = infer_capabilities(ucm.type, _content_dict(ucm))
    unsupported = [c for c in caps if not channel_supports(channel, c)]
    if unsupported:
        steps.append(
            DegradationStep(
                reason="capability",
                from_type=ucm.type,
                to_type="text",
                detail=f'channel "{channel.name}" lacks {", ".join(unsupported)}',
            )
        )
        return _to_text_fallback(ucm)

    if ucm.type == "composite":
        limit = channel.limits.composite_max_depth
        if limit is not None and _composite_depth(ucm) > limit:
            steps.append(
                DegradationStep(
                    reason="depth",
                    from_type="composite",
                    to_type="composite (flattened)",
                    detail=f"depth > {limit}",
                )
            )
            flat_children = _flatten_composite_children(ucm)
            children = [_degrade_inner(c, channel, steps) for c in flat_children]
            return CompositeUCM(
                ucm_version=ucm.ucm_version,
                message_id=ucm.message_id,
                type="composite",
                capabilities_required=ucm.capabilities_required,
                fallback_text=ucm.fallback_text,
                metadata=ucm.metadata,
                content={"children": children},
            )
        children = [_degrade_inner(c, channel, steps) for c in ucm.content.children]
        return CompositeUCM(
            ucm_version=ucm.ucm_version,
            message_id=ucm.message_id,
            type="composite",
            capabilities_required=ucm.capabilities_required,
            fallback_text=ucm.fallback_text,
            metadata=ucm.metadata,
            content={"children": children},
        )

    verdict = validate(ucm.model_dump(mode="json"), channel.name)
    if verdict.ok:
        return ucm

    only_limits = all(i.kind == "limit" for i in verdict.issues)
    if not only_limits:
        steps.append(
            DegradationStep(
                reason="capability",
                from_type=ucm.type,
                to_type="text",
                detail="; ".join(
                    f"{i.kind}@{i.path}: {i.message}" for i in verdict.issues
                ),
            )
        )
        return _to_text_fallback(ucm)

    if ucm.type == "quick_replies":
        return _degrade_quick_replies(ucm, channel, steps)
    if ucm.type == "list":
        return _degrade_list(ucm, channel, steps)
    if ucm.type == "cta_url":
        return _degrade_cta_url(ucm, channel, steps)
    if ucm.type in ("text", "media"):
        return _degrade_text_or_media(ucm, channel, steps)
    return ucm


def _degrade_quick_replies(
    ucm: UCMMessage,
    channel: ChannelProfile,
    steps: list[DegradationStep],
) -> UCMMessage:
    max_btns = channel.limits.quick_replies_max_buttons or 10**9
    title_max = channel.limits.quick_replies_title_max_chars or 10**9

    buttons = [
        {"id": b.id, "title": _truncate(b.title, title_max)}
        for b in ucm.content.buttons
    ]

    if len(buttons) > max_btns:
        if "interactive.list" in channel.capabilities:
            steps.append(
                DegradationStep(
                    reason="limit",
                    from_type="quick_replies",
                    to_type="list",
                    detail=f"{len(buttons)} buttons > {max_btns} — converted to list",
                )
            )
            row_title_max = channel.limits.list_row_title_max_chars or 24
            return ListUCM(
                ucm_version=UCM_VERSION,
                message_id=f"{ucm.message_id}::degraded",
                type="list",
                capabilities_required=["interactive.list"],
                fallback_text=ucm.fallback_text,
                metadata={**ucm.metadata, "degraded_from": "quick_replies"},
                content=ListContent(
                    body=ucm.content.body,
                    button_text=_truncate(
                        "Ver opciones", channel.limits.list_button_text_max_chars or 20
                    ),
                    sections=[
                        ListSection(
                            title=_truncate(
                                "Opciones", row_title_max
                            ),
                            rows=[
                                ListRow(
                                    id=b["id"],
                                    title=_truncate(b["title"], row_title_max),
                                )
                                for b in buttons[
                                    : channel.limits.list_max_rows_total
                                    or len(buttons)
                                ]
                            ],
                        )
                    ],
                ),
            )
        steps.append(
            DegradationStep(
                reason="limit",
                from_type="quick_replies",
                to_type="text",
                detail=(
                    f"{len(buttons)} buttons > {max_btns} — channel cannot list, "
                    "falling back to text"
                ),
            )
        )
        return _to_text_fallback(ucm)

    truncated = any(buttons[i]["title"] != ucm.content.buttons[i].title for i in range(len(buttons)))
    if truncated:
        steps.append(
            DegradationStep(
                reason="limit",
                from_type="quick_replies",
                to_type="quick_replies (truncated)",
                detail=f"button titles truncated to {title_max} chars",
            )
        )
        return ucm.model_copy(
            update={
                "content": ucm.content.model_copy(
                    update={
                        "buttons": [
                            ucm.content.buttons[i].model_copy(update={"title": buttons[i]["title"]})
                            for i in range(len(buttons))
                        ]
                    }
                )
            }
        )
    return ucm


def _degrade_list(
    ucm: UCMMessage,
    channel: ChannelProfile,
    steps: list[DegradationStep],
) -> UCMMessage:
    max_rows = channel.limits.list_max_rows_total or 10**9
    title_max = channel.limits.list_row_title_max_chars or 10**9
    desc_max = channel.limits.list_row_description_max_chars or 10**9
    btn_max = channel.limits.list_button_text_max_chars or 10**9

    new_sections: list[ListSection] = []
    count = 0
    for s in ucm.content.sections:
        new_rows: list[ListRow] = []
        for r in s.rows:
            if count >= max_rows:
                break
            new_rows.append(
                r.model_copy(
                    update={
                        "title": _truncate(r.title, title_max),
                        "description": (
                            _truncate(r.description, desc_max)
                            if r.description is not None
                            else None
                        ),
                    }
                )
            )
            count += 1
        if new_rows:
            new_sections.append(
                s.model_copy(update={"title": _truncate(s.title, title_max), "rows": new_rows})
            )

    total_rows_orig = sum(len(s.rows) for s in ucm.content.sections)
    if total_rows_orig > max_rows:
        steps.append(
            DegradationStep(
                reason="limit",
                from_type="list",
                to_type="list (truncated)",
                detail=f"{total_rows_orig} rows > {max_rows} — truncated",
            )
        )

    new_button = _truncate(ucm.content.button_text, btn_max)
    if new_button != ucm.content.button_text:
        steps.append(
            DegradationStep(
                reason="limit",
                from_type="list",
                to_type="list (truncated button_text)",
                detail=f"button_text truncated to {btn_max} chars",
            )
        )

    return ucm.model_copy(
        update={
            "content": ucm.content.model_copy(
                update={"button_text": new_button, "sections": new_sections}
            )
        }
    )


def _degrade_cta_url(
    ucm: UCMMessage,
    channel: ChannelProfile,
    steps: list[DegradationStep],
) -> UCMMessage:
    max_ = channel.limits.cta_url_button_title_max_chars or 10**9
    new_title = _truncate(ucm.content.button_title, max_)
    if new_title != ucm.content.button_title:
        steps.append(
            DegradationStep(
                reason="limit",
                from_type="cta_url",
                to_type="cta_url (truncated button_title)",
                detail=f"button_title truncated to {max_} chars",
            )
        )
    return ucm.model_copy(
        update={"content": ucm.content.model_copy(update={"button_title": new_title})}
    )


def _degrade_text_or_media(
    ucm: UCMMessage,
    channel: ChannelProfile,
    steps: list[DegradationStep],
) -> UCMMessage:
    max_ = channel.limits.text_body_max_chars or 10**9
    if ucm.type == "text":
        new_body = _truncate(ucm.content.body, max_)
        if new_body != ucm.content.body:
            steps.append(
                DegradationStep(
                    reason="limit",
                    from_type="text",
                    to_type="text (truncated)",
                    detail=f"body truncated to {max_} chars",
                )
            )
        return ucm.model_copy(
            update={"content": ucm.content.model_copy(update={"body": new_body})}
        )
    # media
    if ucm.content.caption is not None:
        new_caption = _truncate(ucm.content.caption, max_)
        if new_caption != ucm.content.caption:
            steps.append(
                DegradationStep(
                    reason="limit",
                    from_type="media",
                    to_type="media (truncated caption)",
                    detail=f"caption truncated to {max_} chars",
                )
            )
        return ucm.model_copy(
            update={"content": ucm.content.model_copy(update={"caption": new_caption})}
        )
    return ucm


def _to_text_fallback(ucm: UCMMessage) -> TextUCM:
    return TextUCM(
        ucm_version=UCM_VERSION,
        message_id=f"{ucm.message_id}::fallback",
        type="text",
        capabilities_required=["text"],
        fallback_text=ucm.fallback_text,
        metadata={**ucm.metadata, "degraded_from": ucm.type},
        content=TextContent(body=ucm.fallback_text, format="plain"),
    )


def _composite_depth(ucm: UCMMessage) -> int:
    if ucm.type != "composite":
        return 0
    m = 0
    for c in ucm.content.children:
        m = max(m, _composite_depth(c))
    return 1 + m


def _flatten_composite_children(ucm: UCMMessage) -> list[UCMMessage]:
    if ucm.type != "composite":
        raise ValueError("flatten on non-composite")
    out: list[UCMMessage] = []
    for child in ucm.content.children:
        if child.type == "composite":
            out.extend(_flatten_composite_children(child))
        else:
            out.append(child)
    return out


def _truncate(s: str, max_len: int) -> str:
    if len(s) <= max_len:
        return s
    if max_len <= 1:
        return s[:max_len]
    return s[: max(1, max_len - 1)] + "…"
